### ISSUE TITLE
[Phase 1] Environment Setup - CUDA, Python & Project Structure

### ISSUE THREAD

**Issue by zordhalo** [zordhalo/photo-organizer#4]
## 🎯 Objective
Set up the development environment with GPU acceleration support, Python dependencies, and project structure for the Construction Photo Analyzer backend.

### 1. GPU Environment Setup
- [ ] Verify NVIDIA GPU availability and driver version
- [ ] Install CUDA Toolkit 11.8 or 12.1
- [ ] Install cuDNN library for deep learning acceleration
- [ ] Verify GPU setup with `nvidia-smi` command
- [ ] Test CUDA availability with PyTorch

### 2. Python Environment
- [ ] Install Python 3.10 or 3.11
- [ ] Create virtual environment: `python -m venv venv`
- [ ] Activate virtual environment
- [ ] Upgrade pip: `pip install --upgrade pip`

### 3. Core Dependencies Installation
- [ ] Install PyTorch with CUDA support:
  ```bash
  pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118
  ```
- [ ] Install FastAPI and Uvicorn:
  ```bash
  pip install fastapi uvicorn[standard]
  ```
- [ ] Install additional dependencies:
  ```bash
  pip install python-multipart pillow pydantic python-dotenv
  ```
- [ ] Install development dependencies:
  ```bash
  pip install pytest pytest-asyncio pytest-cov black flake8
  ```

### 4. Project Structure Setup
- [ ] Create backend directory structure:
  ```
  backend/
  ├── app/
  │   ├── __init__.py
  │   ├── main.py
  │   ├── config.py
  │   ├── api/
  │   │   ├── __init__.py
  │   │   ├── routes.py
  │   │   └── dependencies.py
  │   ├── models/
  │   │   ├── __init__.py
  │   │   └── schemas.py
  │   ├── services/
  │   │   ├── __init__.py
  │   │   └── vision_service.py
  │   └── utils/
  │       ├── __init__.py
  │       └── helpers.py
  ├── tests/
  │   ├── __init__.py
  │   ├── test_api.py
  │   └── test_vision.py
  ├── requirements.txt
  ├── .env.example
  ├── .gitignore
  └── README.md
  ```

### 5. Configuration Files
- [ ] Create `requirements.txt` with all dependencies
- [ ] Create `.env.example` with configuration template:
  ```env
  # Server Configuration
  HOST=0.0.0.0
  PORT=8000
  
  # GPU Configuration
  CUDA_VISIBLE_DEVICES=0
  USE_GPU=true
  
  # Model Configuration
  MODEL_NAME=resnet50
  BATCH_SIZE=8
  
  # API Configuration
  CORS_ORIGINS=http://localhost:3000,http://localhost:5173
  ```
- [ ] Create `.gitignore` for Python/PyTorch projects
- [ ] Create backend README.md with setup instructions

### 6. Verification Tests
- [ ] Create `verify_setup.py` script to test:
  - Python version
  - CUDA availability
  - PyTorch GPU access
  - Required packages installation
  - Import all core dependencies

## 🔧 Verification Script

```python
# verify_setup.py
import sys
import torch
import torchvision
import fastapi
from PIL import Image

def verify_environment():
    print("=" * 50)
    print("Environment Verification")
    print("=" * 50)
    
    # Python version
    print(f"Python Version: {sys.version}")
    
    # PyTorch
    print(f"\nPyTorch Version: {torch.__version__}")
    print(f"Torchvision Version: {torchvision.__version__}")
    
    # CUDA
    cuda_available = torch.cuda.is_available()
    print(f"\nCUDA Available: {cuda_available}")
    
    if cuda_available:
        print(f"CUDA Version: {torch.version.cuda}")
        print(f"GPU Count: {torch.cuda.device_count()}")
        print(f"GPU Name: {torch.cuda.get_device_name(0)}")
        print(f"GPU Memory: {torch.cuda.get_device_properties(0).total_memory / 1e9:.2f} GB")
    
    # Test GPU tensor
    if cuda_available:
        x = torch.rand(5, 3, 224, 224).cuda()
        print(f"\nTest Tensor on GPU: {x.device}")
    
    # FastAPI
    print(f"\nFastAPI Version: {fastapi.__version__}")
    
    print("\n" + "=" * 50)
    print("✅ Environment verification complete!")
    print("=" * 50)

if __name__ == "__main__":
    verify_environment()
```

### Minimum Requirements
- **GPU**: NVIDIA GPU with CUDA Compute Capability 3.5+
- **VRAM**: 4GB+ (8GB recommended)
- **RAM**: 8GB+ (16GB recommended)
- **Storage**: 10GB+ free space
- **OS**: Ubuntu 20.04+, Windows 10/11, macOS (CPU only)

### Recommended Setup
- **GPU**: NVIDIA RTX 3060 or better
- **VRAM**: 8GB+
- **RAM**: 16GB+
- **Storage**: 20GB+ SSD

## ✅ Acceptance Criteria
- [ ] CUDA and GPU drivers properly installed and detected
- [ ] Python 3.10+ environment created and activated
- [ ] All dependencies installed without errors
- [ ] Project structure created with all required directories
- [ ] Configuration files in place
- [ ] `verify_setup.py` runs successfully and confirms GPU availability
- [ ] `pip list` shows all required packages
- [ ] Simple PyTorch GPU test executes successfully

## 📚 References
- [CUDA Installation Guide](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/)
- [PyTorch Get Started](https://pytorch.org/get-started/locally/)
- [FastAPI Documentation](https://fastapi.tiangolo.com/)
- [RunPod GPU Setup](https://www.runpod.io/articles/guides/deploy-fastapi-applications-gpu-cloud)

## ⏱️ Estimated Time
2 hours

## 🔗 Dependencies
None - This is the foundation phase

## 🏷️ Labels
`backend`, `setup`, `infrastructure`, `phase-1`

### For AI Coding Agents:
1. **Verify GPU availability first** - Run `nvidia-smi` before proceeding
2. **Use virtual environment** - Always work within venv to avoid conflicts
3. **Install PyTorch with correct CUDA version** - Match CUDA toolkit version
4. **Test after each major step** - Don't proceed if verification fails
5. **Create all directories at once** - Use mkdir -p for nested structure
6. **Generate requirements.txt from pip freeze** - After confirming all works
7. **Document any issues encountered** - Add to issue comments for future reference
